src/eth/Web3Service.js

```diff
@@ -62,27 +115,107 @@ export default class Web3Service extends PrivateService {
     return new this._web3.eth.Contract(abi, address);
   }
 
+  subscribeNewBlocks(cb) {
+    this._blockSub = this._web3.eth
+      .subscribe('newBlockHeaders')
+      .on('data', blockHeader => {
+        cb(blockHeader);
+      });
+  }
+
+  unsubscribeNewBlocks() {
+    this._blockSub.unsubscribe((err, success) => {
+      if (!success) throw new Error(err);
+      this.get('log').info('Web3 unsubscribing from newBlockHeaders...');
+    });
+  }
+
+  subscribeLog(info, event) {
+    const { address, abi } = info;
+    const res = abi.reduce(
+      (acc, target) => ({
+        ...acc,
+        [target.name]: target
+      }),
+      {}
+    );
+
+    const getTopics = () => {
+      const { sha3 } = this._web3.utils;
+      let topics = [];
+      let name = res[event].name + '(';
+      for (let i in res[event].inputs) {
+        name += res[event].inputs[i].type + ',';
+      }
+      topics[0] = sha3(name.substring(0, name.length - 1) + ')');
+      return topics;
+    };
+    return new Promise((resolve, reject) => {
+      this._eventSub = this._web3.eth.subscribe(
+        'logs',
+        { address, topics: getTopics() },
+        (err, log) => {
```

Comment:
```javascript
    return new Promise((resolve, reject) => {
      const sub = this._web3.eth.subscribe(
        'logs',
	{ address, topics: getTopics() },
        async (err, rawLogData) => {
          if (err) reject(err);
          const log = parseRawLog(rawLogData, res[event]);
          const blockTx = await this._web3.eth.getTransaction(rawLogData.transactionHash);
          const sender = blockTx.from.toLowerCase();
	  if (sender === this.currentAccount()) {
            sub.unsubscribe((err, success) => {
              if (!success) {
                reject(err);
              }
            });
            resolve(log);
          }
        }
      );
    });
```

Reply:
As we discussed during standup, I think it still makes sense to keep the test for matching the event as a callback supplied by the caller, so that `waitForMatchingEvent` can be more general-purpose.